pkg/registry/registry_to_model.go

```diff
@@ -0,0 +1,232 @@
+	if skipRange != "" {
+		out = append(out, property.MustBuildSkipRange(skipRange))
+	}
+
+	replaces, err := b.csv.GetReplaces()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get Replaces from CSV for bundle: %s", err)
+	}
+	for _, ch := range extractChannels(b.Annotations.Channels) {
+		out = append(out, property.MustBuildChannel(ch, replaces))
+	}
+
+	providedGVKs := map[property.GVK]struct{}{}
+	requiredGVKs := map[property.GVKRequired]struct{}{}
+
+	foundPackageProperty := false
+	for i, p := range b.Properties {
+		switch p.Type {
+		case property.TypeGVK:
+			var v property.GVK
+			if err := json.Unmarshal(json.RawMessage(p.Value), &v); err != nil {
+				return nil, property.ParseError{Idx: i, Typ: p.Type, Err: err}
+			}
+			k := property.GVK{Group: v.Group, Kind: v.Kind, Version: v.Version}
+			providedGVKs[k] = struct{}{}
+		case property.TypePackage:
+			foundPackageProperty = true
+			out = append(out, property.Property{
+				Type:  property.TypePackage,
+				Value: json.RawMessage(p.Value),
+			})
+		default:
+			out = append(out, property.Property{
+				Type:  p.Type,
+				Value: json.RawMessage(p.Value),
+			})
+		}
+	}
+
+	for i, p := range b.Dependencies {
+		switch p.Type {
+		case property.TypeGVK:
+			var v property.GVK
+			if err := json.Unmarshal(json.RawMessage(p.Value), &v); err != nil {
+				return nil, property.ParseError{Idx: i, Typ: p.Type, Err: err}
+			}
+			k := property.GVKRequired{Group: v.Group, Kind: v.Kind, Version: v.Version}
+			requiredGVKs[k] = struct{}{}
+		case property.TypePackage:
+			out = append(out, property.Property{
+				Type:  property.TypePackageRequired,
+				Value: json.RawMessage(p.Value),
+			})
+		}
+	}
+
+	version, err := b.Version()
+	if err != nil {
+		return nil, fmt.Errorf("error getting bundle version from CSV %q:%v", b.csv.Name, err)
+	}
+
+	if !foundPackageProperty {
+		out = append(out, property.MustBuildPackage(b.Annotations.PackageName, version))
+	}
+
+	providedApis, err := b.ProvidedAPIs()
+	if err != nil {
+		return nil, fmt.Errorf("error getting Provided APIs for bundle %q:%v", b.Name, err)
+	}
+
+	for p := range providedApis {
+		k := property.GVK{Group: p.Group, Kind: p.Kind, Version: p.Version}
+		if _, ok := providedGVKs[k]; !ok {
+			providedGVKs[k] = struct{}{}
+		}
+	}
+	requiredApis, err := b.RequiredAPIs()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get Required APIs from bundle:%s", err)
+	}
+	for p := range requiredApis {
+		k := property.GVKRequired{Group: p.Group, Kind: p.Kind, Version: p.Version}
+		if _, ok := requiredGVKs[k]; !ok {
+			requiredGVKs[k] = struct{}{}
+		}
+	}
+
+	for p := range providedGVKs {
+		out = append(out, property.MustBuildGVK(p.Group, p.Version, p.Kind))
+	}
+
+	for p := range requiredGVKs {
+		out = append(out, property.MustBuildGVKRequired(p.Group, p.Version, p.Kind))
+	}
+
+	return out, nil
+}
+
+func convertToModelRelatedImages(csv *ClusterServiceVersion) ([]model.RelatedImage, error) {
```

Comment:
This func is the same with what Joe has on his [PR](https://github.com/operator-framework/operator-registry/pull/601/files#diff-485c5555a31cbf01e06186590ac6cf4b5acead26777114cdffeb884ed54ba376R128). It seems it would help with reducing the duplicate code, we should move this func to a command place somewhere to be reused for both PRs.

Reply:
We've discussed this, and since we're dealing with two different types of bundles here (api bundles vs registry bundles) I don't think there's much value in trying to find a common place to put just this function while the rest of the things are unique. Not yet at least.

Reply:
In term of func signature (input, output), they are pretty much identical. Yours accepts `*ClusterServiceVersion` while Joe's accepts a CSV JSON string (which is converted to CSV object later) and both output `([]model.RelatedImage, error)`. They can be just one func with small adjustment on the input side. Either way, this is not a blocker and both of your PRs are in parallel pending so we can take care of this at another time.